### DISCUSSION TITLE
Run command in different shell without exiting current poetry shell.

### DISCUSSION THREAD

**Discussion by real-yfprojects** [python-poetry/poetry#7905] [Q&A] [answered]
I have written a script that runs a specific build workflow (sphinx) for multiple versions of my project. It copies the contents of specific commits into a temporary directory each and runs the build command there. However I ran into a problem with differing dependencies between versions. The build command itself (sphinx) and its extensions (also python packages) are configured in `pyproject.toml`. 

1. I activate the dev env for my usual development tasks `poetry env`
2. I run the build script with `poetry run my-multibuild-script`
3. The build script creates temporary directory (using `tempfile`) and run poetry in set directory using `subprocess`: `poetry run sphinx-build`
4. The build fails because of a missing sphinx-extension set in `conf.py` and `pyproject.toml`.

The complete code can be found [here](https://github.com/real-yfprojects/darglint2/blob/sphinx-docs/docs/sphinx_polyversion.py).

In general when having any poetry shell activated one cannot run commands in a different shell without exiting.

**Comment by radoering** [comment, marked as answer]
> In general when having any poetry shell activated one cannot run commands in a different shell without exiting.

Not sure if I understand your issue, but if you want to execute a poetry command in another virtual environment while being inside a virtual environment of an outer `poetry run` you can try to remove `"VIRTUAL_ENV"` from the `env` you pass to the `subprocess`. That way, the inner poetry command will not use the outer environment.

**Reply by real-yfprojects** [reply]
Thanks! That did the trick, although I had to run `poetry install` in my script before running sphinx with `poetry run`.

**Reply by nealkruis** [reply]
For those coming across this looking for the same thing, here's how I set up the outer environment (though maybe there is a better way?):

```python
import os
import copy
import subprocess

outer_environment = copy.deepcopy(os.environ)
outer_environment.pop("VIRTUAL_ENV", None)

subprocess.Popen(["poetry", "install"], env=outer_environment)
```
